### PR TITLE
adapters: Codify JSON unmarshalling of Big Floats

### PR DESCRIPTION
**Pull request by felder-cl**
Make JSON encoding more orthogonal by creating a new type for the JSON
encoding of big.Floats.

Waiting on #2225

### CODE REVIEW

File: core/adapters/multiply.go [hunk 1]
```diff
@@ -3,36 +3,33 @@
...
 	"chainlink/core/store"
 	"chainlink/core/store/models"
-	"chainlink/core/utils"
 )
 
-// Multiplier represents the number to multiply by in Multiply adapter.
-type Multiplier float64
+// Multiply holds the a number to multiply the given value by.
+type Multiply struct {
+	Times *big.Float `json:"-"`
```
Comment (samphilipd) [inline core/adapters/multiply.go]: What does this '-' do?

File: core/adapters/multiply.go [hunk 2]
```diff
@@ -3,36 +3,33 @@
...
 
-// UnmarshalJSON implements json.Unmarshaler.
-func (m *Multiplier) UnmarshalJSON(input []byte) error {
-	input = utils.RemoveQuotes(input)
-	times, err := strconv.ParseFloat(string(input), 64)
+// UnmarshalJSON implements the json.Unmarshal interface.
+func (ma *Multiply) UnmarshalJSON(buf []byte) error {
+	obj := struct {
+		*Multiply
+		Times *JSONBigFloat `json:"times"`
```
Comment (samphilipd) [inline core/adapters/multiply.go]: Is unmarshalling into this struct covered by tests now?

File: core/adapters/multiply.go (hunk 1 above)
Comment (felder-cl) [reply core/adapters/multiply.go]: It says field is ignored for marshalling JSON.  I added it to signal JSON handling is done elsewhere, but its not required.

**Comment by j16r** [comment]
What makes this more orthogonal?

**Comment by felder-cl** [comment]
> What makes this more orthogonal?

Mainly the naming of the type.  The testing as well was changed to separate JSON decoding from the Multiplier's Perform action.  This is related to [another PR](https://github.com/smartcontractkit/chainlink/pull/2145)

This also changes the conversion from big.Float -> Float -> Big.Float.

The Multiplier's UnmarshalJSON(..) function may not be necessary.

File: core/adapters/multiply_test.go [hunk 3]
```diff
@@ -11,51 +12,64 @@ import (
...
 			input := cltest.NewRunInputWithString(t, test.json)
-			adapter := adapters.Multiply{}
-			jsonErr := json.Unmarshal([]byte(test.params), &adapter)
+			adapter := adapters.Multiply{Times: test.Times}
 			result := adapter.Perform(input, nil)
 
-			if test.jsonError {
-				assert.Error(t, jsonErr)
-			} else if test.errored {
+			if test.errored {
```
Comment (j16r) [inline core/adapters/multiply_test.go]: Would you mind splitting out the one error case here? Logic in tests is a bug bear of mine.

File: core/adapters/multiply.go [hunk 4]
```diff
@@ -3,34 +3,38 @@
...
+}
 
-	*m = Multiplier(times)
+type jsonMultiply struct {
+	Times *JSONBigFloat `json:"times"`
+}
 
-	return nil
+// MarshalJSON implements the json.Marshal interface.
+func (ma *Multiply) MarshalJSON() ([]byte, error) {
```
Comment (j16r) [inline core/adapters/multiply.go]: Why is this necessary?

File: core/adapters/multiply.go [hunk 5]
```diff
@@ -3,34 +3,38 @@
...
+	jsonObj := jsonMultiply{(*JSONBigFloat)(ma.Times)}
+	return json.Marshal(jsonObj)
 }
 
-// Multiply holds the a number to multiply the given value by.
-type Multiply struct {
-	Times *Multiplier `json:"times"`
+// UnmarshalJSON implements the json.Unmarshal interface.
+func (ma *Multiply) UnmarshalJSON(buf []byte) error {
```
Comment (j16r) [inline core/adapters/multiply.go]: As above?

File: core/adapters/json.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package adapters
+
+import (
+	"encoding/json"
+	"math/big"
+)
+
+// JSONBigFloat accepts both string and float JSON values.
+type JSONBigFloat big.Float
```
Comment (j16r) [inline core/adapters/json.go]: We have a bunch of util objects like this in the `utils` package this should probably live along side.

File: core/adapters/multiply.go (hunk 4 above)
Comment (felder-cl) [reply core/adapters/multiply.go]: Its a bit opinionated, it was to keep types to support JSON from bleeding into other parts of the code, which is not a lot.  Putting it another way, are the adapters structures part of the API or fields to be used internally?
